api/utils/keys/yubikey.go

```diff
@@ -0,0 +1,384 @@
+/*
+Copyright 2022 Gravitational, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keys
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"encoding/pem"
+	"io"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/go-piv/piv-go/piv"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api"
+)
+
+const (
+	// PIVCardTypeYubiKey is the PIV card type assigned to yubiKeys.
+	PIVCardTypeYubiKey = "yubikey"
+)
+
+var (
+	// We use slot 9a for Teleport Clients which require `private_key_policy: hardware_key`.
+	pivSlotNoTouch = piv.SlotAuthentication
+	// We use slot 9c for Teleport Clients which require `private_key_policy: hardware_key_touch`.
+	pivSlotWithTouch = piv.SlotSignature
+)
+
+// GetOrGenerateYubiKeyPrivateKey connects to the yubiKey with the given serial number and gets a private key
+// matching the given touch requirement. This private key will either be newly generated or previously
+// generated by a Teleport client and reused.
+func GetOrGenerateYubiKeyPrivateKey(touchRequired bool) (*PrivateKey, error) {
+	// Use the first yubiKey we find.
+	y, err := findYubiKey(0)
```

Comment:
Do we plan on letting the user select keys in the future in some way?

Reply:
Probably not, as this idea was rejected during RFD review. The current goal is to enable PIV login without any additional flags, options, etc.